src/jit/lsra.cpp

```diff
@@ -6653,23 +6656,25 @@ LinearScan::allocateRegisters()
                 }
                 else
 #endif // FEATURE_SIMD
-                if (currentRefPosition->IsActualRef())
+                if (currentRefPosition->RequiresRegister() ||
+                    currentRefPosition->IsRegOptional())
                 {
+                    // Though Lower/Codegen has indicated that it can generate code even if
+                    // no reg is allocated to this ref position, we will make an attempt
+                    // to get a busy reg if it is allocated to a lesser important ref position.
```

Comment:
I have updated the comment to start with "Even if this is marked as optonal,..."

As per my understanding we cannot use `if (allocateReg)` alone here.  Prior to reg-optional changes here is the logic

```
// try to allocate a register
assignedRegister = tryAllocateFreeReg()

if (assignedRegister == REG_NA)
{
if (refType == RefTypeUpperVectorSaveDef)
{
....
}
else if (currentRefPosition->RequiresRegister()
{
   assignedRegister = allocateBusyReg(..)
   ...
}
else
{
...
 }
}
```

As you can see from above tryAllocateFreeReg() was always done even irrespective of RequiresReg() is true or false.  But allocation of a busy reg is conditional on RequiresReg() being true.

allocateReg - this is set to false only those ref positions that are reg-optional, last use and already in spilled state (i.e. needs reload).  When it is set to false, LSRA will not at all attempt to allocate a register (i.e no call made to tryAllocateFreeRegister() and allocateBusyRegister())